service/resource/deployment/resource.go

```diff
@@ -161,47 +161,37 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interf
 	return deployments, nil
 }
 
-// GetCreateState returns the deployments that should be created for this
+// NewUpdatePatch returns the deployments that should be created for this
 // cluster.
-func (r *Resource) GetCreateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error) {
-	currentDeployments, err := toDeployments(currentState)
-	if err != nil {
-		return nil, microerror.Mask(err)
-	}
-	desiredDeployments, err := toDeployments(desiredState)
+func (r *Resource) NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
+	patch := framework.NewPatch()
+
+	deploymentsToCreate, err := r.newCreateChange(ctx, obj, currentState, desiredState)
 	if err != nil {
 		return nil, microerror.Mask(err)
 	}
 
-	var deploymentsToCreate []Deployment
-
-	for _, desiredDeployment := range desiredDeployments {
-		if !existsDeploymentByName(currentDeployments, desiredDeployment.Name) {
-			deploymentsToCreate = append(deploymentsToCreate, desiredDeployment)
-		}
+	if len(deploymentsToCreate) > 0 {
+		patch.SetCreateChange(deploymentsToCreate)
```

Comment:
Why are we using a condition here? I think we should always set the patch to have the unified `NewUpdatePatch ` methods. In this case `newCreateChange` should just make sure to return `nil` in case no deployments should be created.

Reply:
If `newCreateChange` returns a nil we still need the condition. Otherwise `ApplyCreateChange` will be invoked.

Reply:
Yes, and `ApplyCreateChange` can just do nothing in case there is no useful create state given. This is how we did it so far in all of the resource implementations. See for instance https://github.com/giantswarm/kvm-operator/blob/7bf46cd0c5130d104516c854c420d78b99117eef/service/resource/namespace/create.go#L23-L36.

Reply:
I find the debug logging that we get that way useful.

Reply:
> I think we should always set the patch to have the unified NewUpdatePatch methods.

That doesn't make sense to me. What you wrote boils down to this interface:

```
type Resource interface {
	GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error)
	GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error)

	GetCreateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error)
	GetDeleteState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error)
	GetUpdateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error)

	ProcessCreateState(ctx context.Context, obj, createState interface{}) error
	ProcessDeleteState(ctx context.Context, obj, deleteState interface{}) error
	ProcessUpdateState(ctx context.Context, obj, updateState interface{}) error
}
```

Which differs from the old one only by this line:

```
	GetUpdateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, interface{}, interface{}, error)
```

By providing https://github.com/giantswarm/operatorkit/pull/92 this PR I wanted allow the Resource deciding which actions are necessary to reconcile towards the desired state.

Say a resource in order to update has to create and delete something. For example bridge name changes and the resource has to create a new one and delete the old one. To achieve that it sets create and delete patch parts.

Say another resource in order to update it just has to call update api. For example pod definition changes and the resource has to call K8s update api. To achieve that is sets update patch part.

I don't think we can have unified `NewUpdatePatch` methods. Otherwise it wouldn't make sense to provide Patch struct at all.

- - -

I will change according to your request.

Reply:
Done.

Reply:
Why did you do it now after writing a book about not to do it?

Reply:
I wrote a book to make sure we're on the same page.
I did it like you want because I don't care that much and this makes you happy <3

Reply:
Ok, to be on the same page. My goal was not to have fully unified `New*Patch` methods. The patch concept is powerful in that sense that the implementor can choose if create/update/delete state should be computed. In case updating only means updating it is ok to only dispatch an update change. In another case it is also ok to dispatch an update and a create change (or an additional delete change). I only think that we should not use any conditions within the `New*Patch` methods to keep them free of dependencies and logic. I see them only being used to dispatch changes, while they are free to choose which changes (create, update, delete) to dispatch.

Reply:
I wanted to do it other way around. Keep logic in `New*Patch` and remove as much logic from `Apply*Change` as possible. The reason for that is we don't test `Apply*Change` functions, and testing `New*Patch` functions is easy. Testing `Apply*Change` would probably require some third party component like Kubernetes, or etcd, or even whole OS, so it makes sense for me to keeps those things lean.

I understand you like the logs. I think most of them can be in the `New*Patch` as well, but I didn't think much about it yet.

Reply:
IMO there shouldn't be any logic but apply the state using some dependency/client. I do not mind one simple condition though, especially not when this has always the same purpose.